docs: Airflow OpenTelemetry Docs (#15269)
newrelic/docs-website: src/content/docs/more-integrations/open-source-telemetry-integrations/airflow/monitoring-airflow-ot.mdx

---

> title: Monitor Airflow with OpenTelemetry and New Relic 

Reviewer: ```suggestion
title: Monitor Apache Airflow with OpenTelemetry
```

tags:
  - Integrations
  - Open source telemetry integrations
  - OpenTelemetry
  - Airflow
  - Quickstart

> metaDescription: Here is a simple example to setup Airflow OpenTelemetry metrics to send data to New Relic.

Reviewer: ```suggestion
metaDescription: Here is a simple example of setting up Airflow OpenTelemetry metrics to send data to New Relic.
```

redirects:
  - /docs/more-integrations/open-source-telemetry-integrations/opentelemetry/airflow/monitoring-airflow-ot
---

import opentelemetryAirflow01 from 'images/opentelemetry_screenshot_airflow_01.webp'
import opentelemetryAirflow02 from 'images/opentelemetry_screenshot_airflow_02.webp'

> Monitor Airflow with [OpenTelemetry Metrics](https://airflow.apache.org/docs/apache-airflow/stable/administration-and-deployment/logging-monitoring/metrics.html#setup-opentelemetry) by visualizing tasks, operators, and DAG executions as metrics.

Reviewer: ```suggestion
Monitor Apache Airflow data by configuring [OpenTelemetry](https://airflow.apache.org/docs/apache-airflow/stable/administration-and-deployment/logging-monitoring/metrics.html#setup-opentelemetry) to send data to New Relic, where you can visualize tasks, operators, and DAG executions as metrics.
```

<img
  title="Screenshot showing sample Airflow DAG runs in New Relic"
  alt="Screenshot showing sample Airflow DAG runs in New Relic"
  src={opentelemetryAirflow01}
/>

## Prerequisites [#prerequisites]

> You need to first install the Airflow package with OpenTelemetry using the `otel` extra. This can be done in one of several ways, depending on how you deploy Airflow:

Reviewer: ```suggestion
Before enabling OpenTelemetry in Apache Airflow, you'll need to install the Airflow package with the `otel` extra. The installation method depends on your Airflow deployment approach:
```

> ### Installing from PyPi [#install-pypi]

Reviewer: ```suggestion
### Option 1: Installing from PyPi [#install-pypi]
```

1. Follow the installation instructions from [Airflow's Documentation](https://airflow.apache.org/docs/apache-airflow/stable/installation/installing-from-pypi.html).
2. When installing with pip, add the `otel` extra to the command.
  a. eg. `pip install "apache-airflow[otel]"`

> ### Installing from Docker [#install-docker]

Reviewer: ```suggestion
### Option 2: Installing from Docker [#install-docker]
```

> 1. Setup the [Airflow Docker image](https://airflow.apache.org/docs/docker-stack/index.html) using instructions from [Airflow's Documentation](https://airflow.apache.org/docs/docker-stack/index.html).

Reviewer: ```suggestion
1. Set up the Airflow Docker image using instructions from [Airflow's documentation](https://airflow.apache.org/docs/docker-stack/index.html).
```

> 2. Extend the prebuilt image by using a Dockerfile to install the `otel` extra. You can replace the latest tag with your desired version of the Airflow image.

Reviewer: ```suggestion
2. Extend the pre-built Docker image by using a Dockerfile to install the `otel` extra. You can replace the latest tag with your desired version of the image.
```

```
FROM apache/airflow:latest

> RUN pip install --no-cache-dir "apache-airflow[otel]==${AIRFLOW_VERSION}"

Reviewer: @TimPansino Do they need to replace `AIRFLOW_VERSION`, or `{AIRFLOW_VERSION}`? Can it also be `latest`?

Author: It can't be latest, they've got really complex dependencies that initially get installed with a constraints file. Has to match the installed version.

Conveniently `AIRFLOW_VERSION` is baked into the image as an environment variable so there's no need to replace it.

Author: I made a change to this section, see if that's more clear.

```

## Configuration [#configuration]

> You need to configure the Airflow OpenTelemetry metrics to send data to an [OpenTelemetry Collector](/docs/more-integrations/open-source-telemetry-integrations/opentelemetry/collector/opentelemetry-collector-intro/) that reports data to a New Relic [OTLP Endpoint](/docs/more-integrations/open-source-telemetry-integrations/opentelemetry/opentelemetry-setup/#note-endpoints) using a <InlinePopover type="licenseKey" />.

Reviewer: ```suggestion
To send Airflow metrics to New Relic, configure the OpenTelemetry metrics to export data to an [OpenTelemetry collector](/docs/more-integrations/open-source-telemetry-integrations/opentelemetry/collector/opentelemetry-collector-intro/), which will then forward the data to a New Relic [OTLP endpoint](/docs/more-integrations/open-source-telemetry-integrations/opentelemetry/opentelemetry-setup/#note-endpoints) using a <InlinePopover type="licenseKey" />.
```

<Callout variant="important">

>   Airflow does not currently supporting sending OpenTelemetry data with authentication headers, making the OpenTelemetry Collector a requirement to authenticate with New Relic.

Reviewer: ```suggestion
  Due to Airflow's current lack of support for sending OpenTelemetry data with authentication headers, the OpenTelemetry collector is essential for authenticating with New Relic.
```

</Callout>

> ### OpenTelemetry Collector Configuration [#configuration-collector]

Reviewer: ```suggestion
### Configure the OpenTelemetry collector [#configuration-collector]
```

> 1. Follow the [basic collector example](/docs/more-integrations/open-source-telemetry-integrations/opentelemetry/collector/opentelemetry-collector-basic/) to setup your OpenTelemetry Collector.

Reviewer: ```suggestion
1. Follow the [basic collector example](/docs/more-integrations/open-source-telemetry-integrations/opentelemetry/collector/opentelemetry-collector-basic/) to set up your OpenTelemetry collector.
```

> 2. Configure it with your appropriate OTLP Endpoint. For example, `https://otlp.nr-data.net:4317`.

Reviewer: ```suggestion
2. Configure the collector with your appropriate OTLP endpoint, such as `https://otlp.nr-data.net:4317`.
```

> 3. For authentication, add your <InlinePopover type="licenseKey" /> to the environment variable `NEW_RELIC_LICENSE_KEY` so that it populates the `api-key` header.

Reviewer: ```suggestion
3. For authentication, add your <InlinePopover type="licenseKey" /> to the environment variable `NEW_RELIC_LICENSE_KEY` so that it populates the `api-key` header.
```

Author: This looks like it was supposed to be a suggested change but I see no diff, was there something you wanted edited here?

> 4. Ensure port 4318 on the collector is open and available to the running Airflow instance. (For docker, you may need to use a [docker network](https://docs.docker.com/network/).)

Reviewer: ```suggestion
4. Use port `4318` on the collector to run the Airflow instance. For Docker, you may need to use a [Docker network](https://docs.docker.com/network/).
```

Author: I like your formatting better but the content is wrong, flagging this to edit myself.

Author: ```suggestion
4. Ensure port 4318 on the collector is reachable from the running Airflow instance. (For docker, you may need to use a [docker network](https://docs.docker.com/network/).)
```

Perhaps this? I can't quite seem to word it in a way that sounds clear.

The idea is that the most common issue is networking, where the Airflow container can't reach the collector. The port needs to be open on the collector which requires the config file to be set up right, and for the docker configuration to expose or bind that port. I'm really not sure how best to word this.

> 5. Launch your collector.

Reviewer: ```suggestion
5. Launch the collector.
```

> ### Airflow Metrics Configuration [#configuration-airflow]

Reviewer: ```suggestion
### Configure Airflow metrics [#configuration-airflow]
```

> Airflow sends metrics using OTLP over HTTP. To send this data to our OpenTelemetry Collector it will need to be sent to port 4318. Airflow has multiple methods of [setting configuration options](https://airflow.apache.org/docs/apache-airflow/stable/howto/set-config.html).

Reviewer: ```suggestion
Airflow sends metrics using OTLP over HTTP. To send this data to the OpenTelemetry collector, you'll need to use port `4318`. Airflow has multiple methods of [setting configuration options](https://airflow.apache.org/docs/apache-airflow/stable/howto/set-config.html).
```

Reviewer: Actually, can we remove this? It's implied in the previous step that they already set the port to 4318.

Author: I wanted to make this crystal clear because the docs for Airflow and Airflow's default setting are straight up wrong and say to use port 8889 which is likely copy paste error from prometheus metrics.

I was planning to PR that change but I'm not sure if I feel like navigating multiple repos and Apache's review process. We'll see if I get around to it.

Author: I reformatted this to be shorted, see if that's better.

</Callout>

Choose one of the following methods to set the required options for Airflow.

1. Set the required options in the `airflow.cfg` file.

```
[metrics]
otel_on = True
otel_host = localhost
otel_port = 4318
otel_ssl_active = False
```

2. Or, set the required options as environment variables.

```
export AIRFLOW__METRICS__OTEL_ON=True
export AIRFLOW__METRICS__OTEL_HOST=localhost
export AIRFLOW__METRICS__OTEL_PORT=4318
export AIRFLOW__METRICS__OTEL_SSL_ACTIVE=False
```

<Callout variant="tip">
  Airflow has [additional settings](https://airflow.apache.org/docs/apache-airflow/stable/administration-and-deployment/logging-monitoring/metrics.html#setup-opentelemetry) for metrics that may be useful. This includes the ability to [rename metrics](https://airflow.apache.org/docs/apache-airflow/stable/administration-and-deployment/logging-monitoring/metrics.html#rename-metrics) before sending, which is helpful if metric names exceed the 63 byte limit for OpenTelemetry.
</Callout>

> ## Validation [#validation]

Reviewer: ```suggestion
## Valide data is sent to New Relic [#validation]
```

Author: Typo here, flagging for edit

> Run a DAG or pipeline to see Airflow data in New Relic.

Reviewer: ```suggestion
To confirm New Relic is collecting your Airflow data, run a DAG or pipeline:
```

> 1. Log into Airflow.

Reviewer: ```suggestion
1. Login to Airflow.
```

2. Click the run button on one of the existing tutorial DAGs, or your own.
3. Wait for the pipeline to finish running.
4. Go to **[one.newrelic.com > All capabilities](https://one.newrelic.com/all-capabilities) > APM & services > Services - OpenTelemetry > Airflow**.
5. Click **Metrics Explorer** to visualize metrics for pipeline executions.

## Building dashboards [#building-dashboards]

With Airflow metrics, you can build dashboards around individual pipelines, overall performance, or view a comparison between different pipelines. Click here to learn more about [querying your metrics](/docs/data-apis/understand-data/metric-data/query-metric-data-type/).

> Use the following query to retrieve a list of all reported metrics for Airflow. Be sure to change the limit if your metric names exceed it.

Reviewer: ```suggestion
This query retrieves a list of all reported metrics for Airflow:
```

```
SELECT uniques(metricName) FROM Metric WHERE entity.name = 'Airflow' AND metricName LIKE 'airflow.%' SINCE 30 MINUTES AGO LIMIT 100
```

> 

Reviewer: ```suggestion

Make sure to change the limit (`100`) if your metric names exceed it.

```

> This query shows a comparison of different completion times for successful runs of different DAGs.

Reviewer: ```suggestion
This query shows a comparison of different completion times for successful runs of different DAGs:
```

```
SELECT latest(airflow.dagrun.duration.success) FROM Metric FACET dag_id WHERE entity.name = 'Airflow' SINCE 30 minutes AGO TIMESERIES
```

<img
  title="Screenshot showing sample Airflow DAG runs in New Relic"
  alt="Screenshot showing sample Airflow DAG runs in New Relic"
  src={opentelemetryAirflow01}
/>

> This query shows counts of failed DAG runs, which can be used to build alerts for critical pipelines.

Reviewer: ```suggestion
This query shows counts of failed DAG runs, which can be used to build alerts for critical pipelines:
```

---
title: Monitor Apache Airflow with OpenTelemetry
tags:
  - Integrations
  - Open source telemetry integrations
  - OpenTelemetry
  - Airflow
  - Quickstart

> metaDescription: Monitor Airflow data via OpenTelemetry to New Relic.

Reviewer: ```suggestion
metaDescription: Monitor Airflow data with New Relic using OpenTelemetry.
```
